MoreLinq/ScanRight.cs

```diff
@@ -0,0 +1,115 @@
+        /// </param>
+        /// <returns>The scanned sequence.</returns>
+        /// <example>
+        /// <code>
+        /// var result = Enumerable.Range(1, 5).Select(i => i.ToString()).ScanRight((a, b) => string.Format("({0}/{1})", a, b));
+        /// </code>
+        /// The <c>result</c> variable will contain <c>[ "(1+(2+(3+(4+5))))", "(2+(3+(4+5)))", "(3+(4+5))", "(4+5)", "5" ]</c>.
+        /// </example>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+
+        public static IEnumerable<TSource> ScanRight<TSource>(this IEnumerable<TSource> source, Func<TSource, TSource, TSource> func)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (func == null) throw new ArgumentNullException(nameof(func));
+
+            return _(); IEnumerable<TSource> _()
+            {
+                var list = (source as IList<TSource>) ?? source.ToList();
+
+                if (list.Count == 0)
+                    yield break;
+
+                foreach (var item in ScanRightImpl(list, list.Last(), func, list.Count - 1))
+                    yield return item;
+            }
+        }
+
+        /// <summary>
+        /// Peforms a right-associative scan (inclusive prefix) on a sequence of elements.
+        /// The specified seed value is used as the initial accumulator value.
+        /// This operator is the right-associative version of the 
+        /// <see cref="MoreEnumerable.Scan{TSource, TState}(IEnumerable{TSource}, TState, Func{TState, TSource, TState})"/> LINQ operator.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <typeparam name="TAccumulate">The type of the accumulator value.</typeparam>
+        /// <param name="source">Source sequence.</param>
+        /// <param name="seed">The initial accumulator value.</param>
+        /// <param name="func">A right-associative accumulator function to be invoked on each element.</param>
+        /// <returns>The scanned sequence.</returns>
+        /// <example>
+        /// <code>
+        /// var result = Enumerable.Range(1, 4).ScanRight("5", (a, b) => string.Format("({0}/{1})", a, b));
+        /// </code>
+        /// The <c>result</c> variable will contain <c>[ "(1+(2+(3+(4+5))))", "(2+(3+(4+5)))", "(3+(4+5))", "(4+5)", "5" ]</c>.
+        /// </example>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+
+        public static IEnumerable<TAccumulate> ScanRight<TSource, TAccumulate>(this IEnumerable<TSource> source, TAccumulate seed, Func<TSource, TAccumulate, TAccumulate> func)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (func == null) throw new ArgumentNullException(nameof(func));
+
+            return _(); IEnumerable<TAccumulate> _()
```

Comment:
> Now it is more verbose (more lines) and a bit harder to understand.

I don't think there's more lines. There's a new data structure.

> Why we can't use tuples here?

Because as [I commented in the code, we can once we drop .NET Framework 3.5 as a target](https://github.com/morelinq/MoreLINQ/commit/2656082881c2897e504c8a42b08a13b27f1dc8e3#diff-d45d5bd92cdeef152abc5d8a68122510R89).